File: frontend/src/lib/api/agent-specs.ts
import { type IdParam, type RequestQueryValue, requestV2, toPathSegment } from "../api-client";
import type {
  AgentSpecDraftCreateInput,
  AgentSpecDraftUpdateInput,
  AgentSpecListRead,
  AgentSpecRead,
  StudioSpecListParams,
} from "../types/studio";

function agentSpecPath(specId: IdParam): string {
  return `/agent-specs/${toPathSegment(specId)}`;
}

function toQueryRecord<T extends object>(
  params?: T,
): Record<string, RequestQueryValue> | undefined {
  return params as Record<string, RequestQueryValue> | undefined;
}

export function listAgentSpecs(
  params?: StudioSpecListParams,
  signal?: AbortSignal,
): Promise<AgentSpecListRead> {
  return requestV2<AgentSpecListRead>("/agent-specs", {
    query: toQueryRecord(params),
    signal,
  });
}

export function getAgentSpec(specId: IdParam, signal?: AbortSignal): Promise<AgentSpecRead> {
  return requestV2<AgentSpecRead>(agentSpecPath(specId), { signal });
}

export function createAgentSpec(
  payload: AgentSpecDraftCreateInput,
  signal?: AbortSignal,
): Promise<AgentSpecRead> {
  return requestV2<AgentSpecRead>("/agent-specs", {
    body: payload,
    method: "POST",
    signal,
  });
}

export function updateAgentSpec(
  specId: IdParam,
  payload: AgentSpecDraftUpdateInput,
  signal?: AbortSignal,
): Promise<AgentSpecRead> {
  return requestV2<AgentSpecRead>(agentSpecPath(specId), {
    body: payload,
    method: "PATCH",
    signal,
  });
}

export function activateAgentSpec(specId: IdParam, signal?: AbortSignal): Promise<AgentSpecRead> {
  return requestV2<AgentSpecRead>(`${agentSpecPath(specId)}/activate`, {
    method: "POST",
    signal,
  });
}

export function deprecateAgentSpec(specId: IdParam, signal?: AbortSignal): Promise<AgentSpecRead> {
  return requestV2<AgentSpecRead>(`${agentSpecPath(specId)}/deprecate`, {
    method: "POST",
    signal,
  });
}

export function archiveAgentSpec(specId: IdParam, signal?: AbortSignal): Promise<AgentSpecRead> {
  return requestV2<AgentSpecRead>(`${agentSpecPath(specId)}/archive`, {
    method: "POST",
    signal,
  });
}

export const agentSpecsApi = {
  list: listAgentSpecs,
  get: getAgentSpec,
  create: createAgentSpec,
  update: updateAgentSpec,
  activate: activateAgentSpec,
  deprecate: deprecateAgentSpec,
  archive: archiveAgentSpec,
} as const;

File: backend/app/api/agent_specs.py
from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, Query, status

from app.api.dependencies import get_agent_spec_service
from app.schemas.runtime import SpecLifecycleStatus, SpecOrigin
from app.schemas.studio import (
    AgentSpecDraftCreate,
    AgentSpecDraftUpdate,
    AgentSpecListRead,
    AgentSpecRead,
)
from app.services.agent_spec_service import AgentSpecService

router = APIRouter(prefix="/agent-specs", tags=["agent-specs"])


@router.get("", response_model=AgentSpecListRead)
def list_agent_specs(
    service: Annotated[AgentSpecService, Depends(get_agent_spec_service)],
    origin: Annotated[SpecOrigin | None, Query()] = None,
    status_filter: Annotated[SpecLifecycleStatus | None, Query(alias="status")] = None,
) -> AgentSpecListRead:
    return service.list_specs(origin=origin, status_filter=status_filter)


@router.post("", response_model=AgentSpecRead, status_code=status.HTTP_201_CREATED)
def create_agent_spec_draft(
    payload: AgentSpecDraftCreate,
    service: Annotated[AgentSpecService, Depends(get_agent_spec_service)],
) -> AgentSpecRead:
    return service.create_draft(payload)


@router.get("/{spec_id}", response_model=AgentSpecRead)
def get_agent_spec(
    spec_id: int,
    service: Annotated[AgentSpecService, Depends(get_agent_spec_service)],
) -> AgentSpecRead:
    return service.get_spec(spec_id)


@router.patch("/{spec_id}", response_model=AgentSpecRead)
def update_agent_spec_draft(
    spec_id: int,
    payload: AgentSpecDraftUpdate,
    service: Annotated[AgentSpecService, Depends(get_agent_spec_service)],
) -> AgentSpecRead:
    return service.update_draft(spec_id, payload)


@router.post("/{spec_id}/activate", response_model=AgentSpecRead)
def activate_agent_spec(
    spec_id: int,
    service: Annotated[AgentSpecService, Depends(get_agent_spec_service)],
) -> AgentSpecRead:
    return service.activate(spec_id)


@router.post("/{spec_id}/deprecate", response_model=AgentSpecRead)
def deprecate_agent_spec(
    spec_id: int,
    service: Annotated[AgentSpecService, Depends(get_agent_spec_service)],
) -> AgentSpecRead:
    return service.deprecate(spec_id)


@router.post("/{spec_id}/archive", response_model=AgentSpecRead)
def archive_agent_spec(
    spec_id: int,
    service: Annotated[AgentSpecService, Depends(get_agent_spec_service)],
) -> AgentSpecRead:
    return service.archive(spec_id)
